File: agent_skills/literature_grounding.py
"""Agent skill: Literature grounding for biomarker genes.

Queries PubMed E-utilities API (or mocks the query in test mode) and
optionally synthesises findings with Claude for confidence classification.
"""

from __future__ import annotations

import logging
from typing import Any

logger = logging.getLogger(__name__)

PUBMED_ESEARCH_URL = "https://eutils.ncbi.nlm.nih.gov/entrez/eutils/esearch.fcgi"
PUBMED_EFETCH_URL = "https://eutils.ncbi.nlm.nih.gov/entrez/eutils/efetch.fcgi"


class LiteratureGroundingSkill:
    """Ground biomarker genes in published literature via PubMed."""

    def __init__(self, http_client=None, llm_client=None) -> None:
        self.http_client = http_client
        self.llm_client = llm_client

    async def _pubmed_search(self, gene: str, context: str) -> dict:
        """Search PubMed for a gene in a given context.

        Returns a dict with ``pmid_count``, ``top_pmids``, and ``query``.
        """
        query = f"{gene} AND {context}"
        if self.http_client is None:
            return {
                "query": query,
                "pmid_count": 0,
                "top_pmids": [],
                "source": "mock",
            }

        try:
            response = await self.http_client.get(
                PUBMED_ESEARCH_URL,
                params={
                    "db": "pubmed",
                    "term": query,
                    "retmax": 5,
                    "retmode": "json",
                },
            )
            data = response.json()
            result = data.get("esearchresult", {})
            return {
                "query": query,
                "pmid_count": int(result.get("count", 0)),
                "top_pmids": result.get("idlist", []),
                "source": "pubmed",
            }
        except Exception as exc:
            logger.warning("PubMed search failed for %s: %s", gene, exc)
            return {
                "query": query,
                "pmid_count": 0,
                "top_pmids": [],
                "source": "error",
                "error": str(exc),
            }

    async def _synthesize(self, gene: str, context: str, literature: dict) -> dict:
        """Use an LLM client to synthesise literature findings."""
        if self.llm_client is None:
            pmid_count = literature.get("pmid_count", 0)
            if pmid_count >= 10:
                confidence = "high"
            elif pmid_count >= 3:
                confidence = "medium"
            else:
                confidence = "low"
            return {
                "gene": gene,
                "context": context,
                "confidence": confidence,
                "summary": f"{gene} has {pmid_count} PubMed references in {context} context.",
                "source": "heuristic",
            }

        try:
            prompt = (
                f"Summarise the role of {gene} in {context} based on "
                f"{literature.get('pmid_count', 0)} PubMed references. "
                f"Classify confidence as high/medium/low."
            )
            response = await self.llm_client.messages.create(
                model="claude-sonnet-4-20250514",
                max_tokens=300,
                messages=[{"role": "user", "content": prompt}],
            )
            text = response.content[0].text
            return {
                "gene": gene,
                "context": context,
                "summary": text,
                "confidence": "medium",
                "source": "llm",
            }
        except Exception as exc:
            logger.warning("LLM synthesis failed for %s: %s", gene, exc)
            return {
                "gene": gene,
                "context": context,
                "confidence": "low",
                "summary": f"LLM synthesis failed: {exc}",
                "source": "error",
            }

    async def run(
        self,
        genes: list[str],
        context: str = "MSI",
    ) -> dict:
        """Query PubMed for each gene, synthesise, and return grounding report."""
        results: list[dict[str, Any]] = []

        for gene in genes:
            literature = await self._pubmed_search(gene, context)
            synthesis = await self._synthesize(gene, context, literature)
            results.append(
                {
                    "gene": gene,
                    "literature": literature,
                    "synthesis": synthesis,
                }
            )

        # Aggregate confidence
        confidences = [r["synthesis"]["confidence"] for r in results]
        high_count = confidences.count("high")
        med_count = confidences.count("medium")
        total = len(confidences)

        if total == 0:
            overall = "none"
        elif high_count / total >= 0.5:
            overall = "high"
        elif (high_count + med_count) / total >= 0.5:
            overall = "medium"
        else:
            overall = "low"

        return {
            "genes_queried": genes,
            "context": context,
            "results": results,
            "overall_confidence": overall,
            "n_genes": len(genes),
            "n_high_confidence": high_count,
            "n_medium_confidence": med_count,
            "n_low_confidence": total - high_count - med_count,
        }

File: web/src/lib/agent-skills/literature-grounding.ts
/**
 * Agent skill: Literature grounding for biomarker genes.
 * Migrated from agent_skills/literature_grounding.py.
 *
 * Queries PubMed E-utilities API and optionally synthesizes with Claude.
 */

import Anthropic from "@anthropic-ai/sdk";

const PUBMED_ESEARCH_URL =
  "https://eutils.ncbi.nlm.nih.gov/entrez/eutils/esearch.fcgi";

interface LiteratureResult {
  query: string;
  pmid_count: number;
  top_pmids: string[];
  source: string;
  error?: string;
}

interface SynthesisResult {
  gene: string;
  context: string;
  confidence: "high" | "medium" | "low";
  summary: string;
  source: string;
}

async function pubmedSearch(
  gene: string,
  context: string,
): Promise<LiteratureResult> {
  const query = `${gene} AND ${context}`;
  try {
    const url = new URL(PUBMED_ESEARCH_URL);
    url.searchParams.set("db", "pubmed");
    url.searchParams.set("term", query);
    url.searchParams.set("retmax", "5");
    url.searchParams.set("retmode", "json");

    const res = await fetch(url.toString());
    const data = await res.json();
    const result = data.esearchresult ?? {};
    return {
      query,
      pmid_count: parseInt(result.count ?? "0", 10),
      top_pmids: result.idlist ?? [],
      source: "pubmed",
    };
  } catch (err) {
    return {
      query,
      pmid_count: 0,
      top_pmids: [],
      source: "error",
      error: err instanceof Error ? err.message : String(err),
    };
  }
}

async function synthesize(
  gene: string,
  context: string,
  literature: LiteratureResult,
  useLLM: boolean,
): Promise<SynthesisResult> {
  const pmidCount = literature.pmid_count;

  if (!useLLM) {
    let confidence: "high" | "medium" | "low";
    if (pmidCount >= 10) confidence = "high";
    else if (pmidCount >= 3) confidence = "medium";
    else confidence = "low";

    return {
      gene,
      context,
      confidence,
      summary: `${gene} has ${pmidCount} PubMed references in ${context} context.`,
      source: "heuristic",
    };
  }

  try {
    const client = new Anthropic();
    const message = await client.messages.create({
      model: "claude-sonnet-4-6",
      max_tokens: 300,
      messages: [
        {
          role: "user",
          content: `Summarise the role of ${gene} in ${context} based on ${pmidCount} PubMed references. Classify confidence as high/medium/low.`,
        },
      ],
    });
    const text =
      message.content[0].type === "text" ? message.content[0].text : "";
    return {
      gene,
      context,
      confidence: "medium",
      summary: text,
      source: "llm",
    };
  } catch {
    return {
      gene,
      context,
      confidence: "low",
      summary: `LLM synthesis failed for ${gene}`,
      source: "error",
    };
  }
}

export async function runLiteratureGrounding(
  genes: string[],
  context: string = "MSI",
  useLLM: boolean = false,
): Promise<Record<string, unknown>> {
  const results: Array<Record<string, unknown>> = [];

  for (const gene of genes) {
    const literature = await pubmedSearch(gene, context);
    const synthesis = await synthesize(gene, context, literature, useLLM);
    results.push({ gene, literature, synthesis });
  }

  const confidences = results.map(
    (r) => (r.synthesis as SynthesisResult).confidence,
  );
  const highCount = confidences.filter((c) => c === "high").length;
  const medCount = confidences.filter((c) => c === "medium").length;
  const total = confidences.length;

  let overall: string;
  if (total === 0) overall = "none";
  else if (highCount / total >= 0.5) overall = "high";
  else if ((highCount + medCount) / total >= 0.5) overall = "medium";
  else overall = "low";

  return {
    genes_queried: genes,
    context,
    results,
    overall_confidence: overall,
    n_genes: genes.length,
    n_high_confidence: highCount,
    n_medium_confidence: medCount,
    n_low_confidence: total - highCount - medCount,
  };
}
